src/graph/sampler.cc

```diff
@@ -988,415 +988,769 @@ std::vector<dgl_id_t> Global2Local(const std::vector<size_t> &ids,
-    neg_unchanged = neg_dst_data;
-    neg_changed = neg_src_data;
-  } else {
-    unchanged = src_data;
-    neg_unchanged = neg_src_data;
-    neg_changed = neg_dst_data;
+inline SubgraphRef ConvertRef(const NegSubgraph &subg) {
+    return SubgraphRef(std::shared_ptr<Subgraph>(new NegSubgraph(subg)));
+}
+
+}  // namespace
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_GetNegEdgeExistence")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+  SubgraphRef g = args[0];
+  auto gptr = std::dynamic_pointer_cast<NegSubgraph>(g.sptr());
+  *rv = gptr->exist;
+});
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_GetEdgeSubgraphHead")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+  SubgraphRef g = args[0];
+  auto gptr = std::dynamic_pointer_cast<NegSubgraph>(g.sptr());
+  *rv = gptr->head_nid;
+});
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_GetEdgeSubgraphTail")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+  SubgraphRef g = args[0];
+  auto gptr = std::dynamic_pointer_cast<NegSubgraph>(g.sptr());
+  *rv = gptr->tail_nid;
+});
+
+class EdgeSamplerObject: public Object {
+public:
+  EdgeSamplerObject(const GraphPtr gptr,
+                    IdArray seed_edges,
+                    const int64_t batch_size,
+                    const int64_t num_workers,
+                    const std::string neg_mode,
+                    const int64_t neg_sample_size,
+                    const bool exclude_positive,
+                    const bool check_false_neg,
+                    IdArray relations) {
+    gptr_ = gptr;
+    seed_edges_ = seed_edges;
+    relations_ = relations;
+
+    batch_size_ = batch_size;
+    num_workers_ = num_workers;
+    neg_mode_ = neg_mode;
+    neg_sample_size_ = neg_sample_size;
+    exclude_positive_ = exclude_positive;
+    check_false_neg_ = check_false_neg;
   }
 
-  std::unordered_map<dgl_id_t, dgl_id_t> neg_map;
-  std::vector<dgl_id_t> local_pos_vids;
-  local_pos_vids.reserve(num_pos_edges);
-
-  dgl_id_t curr_eid = 0;
-  std::vector<size_t> neg_vids;
-  neg_vids.reserve(neg_sample_size);
-  // If we don't exclude positive edges, we are actually sampling more than
-  // the total number of nodes in the graph.
-  if (!exclude_positive && neg_sample_size >= num_tot_nodes) {
-    // We add all nodes as negative nodes.
-    for (int64_t i = 0; i < num_tot_nodes; i++) {
-      neg_vids.push_back(i);
-      neg_map[i] = i;
+  ~EdgeSamplerObject() {}
+
+  virtual void fetch(DGLRetValue* rv) = 0;
+
+protected:
+  virtual void randomSample(size_t set_size, size_t num, std::vector<size_t>* out) = 0;
+  virtual void randomSample(size_t set_size, size_t num, const std::vector<size_t> &exclude,
+                    std::vector<size_t>* out) = 0;
+
+  NegSubgraph genNegEdgeSubgraph(const Subgraph &pos_subg,
+                                 const std::string &neg_mode,
+                                 int64_t neg_sample_size,
+                                 bool exclude_positive,
+                                 bool check_false_neg) {
```

Comment:
can you move the implementation of `genNegEdgeSubgraph` so that diff can work correctly to see what is changed and what isn't?